decentraland/unity-explorer
# fix: layered animations

## What does this PR change?

Dependan of AB conversions from: https://github.com/decentraland/asset-bundle-converter/pull/99

Fixes #1087 
Fixes #1264 
Fixes #1099 
Fixes #1102

The SDK supports more than one animations playing simultaneously, each one with different weight values. In order to support it we make a new layer for every anim state in the Animator, so we can apply different weighting values.

Concerns:
- So explicit transitions between states will not exist anymore, since each state stays in its own layer.
- The Base Layer remains empty. Could be useful somehow?
- Performance impact?

Here is how an animator looks the shark that contains two clips: swim, bite.

One layer for each SDK state:

<img width="333" alt="Screenshot 2024-07-04 at 17 04 20" src="https://github.com/decentraland/unity-explorer/assets/56365551/4618c632-305e-40ea-bdeb-6dcb720dc875">


Adds 3 params for each SDK state:
- Trigger: enabled when the animation should start playing.
- Loop: should the state keep on loop?
- Enabled: is the animation playing? When is set to false then it will transition to the empty-state so stops.

<img width="334" alt="Screenshot 2024-07-04 at 17 04 32" src="https://github.com/decentraland/unity-explorer/assets/56365551/540b6d3f-ff55-46ec-acba-51c01981e568">


Here is how states look in a layer:

<img width="406" alt="Screenshot 2024-07-04 at 17 04 52" src="https://github.com/decentraland/unity-explorer/assets/56365551/7a5b443e-35d9-4da6-8457-f01b56b74ab3">


Here are the transitions:

<img width="1378" alt="Screenshot 2024-07-04 at 17 05 13" src="https://github.com/decentraland/unity-explorer/assets/56365551/e1ba0bbd-ccc1-479c-91fb-bcb128ab1a17">

<img width="1379" alt="Screenshot 2024-07-04 at 17 05 25" src="https://github.com/decentraland/unity-explorer/assets/56365551/d8eb7876-d967-46e7-9468-fc428955e745">

<img width="1377" alt="Screenshot 2024-07-04 at 17 05 35" src="https://github.com/decentraland/unity-explorer/assets/56365551/69677824-58fa-4cae-8d81-a9530049a618">

<img width="1378" alt="Screenshot 2024-07-04 at 17 05 47" src="https://github.com/decentraland/unity-explorer/assets/56365551/ef73af6d-9fef-4231-b890-3ae19ca5b3ac">


## How to test the changes?

Affects all animations so an intensive testing should be done.
Before starting, make sure to clear asset bundle cache.
Some of the scenes in goerli that contain animations:

73,-8: the shark should swim by default and bite whenever you click it. It should never stop swimming

80,-1:
- avatar emotes by clicking the cubes and the snowpile
- elf dancing by default

77,-6: stand on green tiles. Should play animations on the avatar

76,-3: guy dancing

74,-9: two skeleton kings should walk, turn and play animation when you are near

78,-1: zombies should walk and attack you when near

72,-9: the dogs should be animated

76,-9: the tree should shake once after you interact with it. The birds should be animated

75,-7 & 76,-7: the avatar swap should play the running animation when you move and stop when you stand still

## Our Code Review Standards

https://github.com/decentraland/unity-renderer/blob/master/docs/code-review-standards.md
